johandry/terranova
# [Feature] Logger

**Community Note**

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Ensure you have added or ran the appropriate tests for this pull request
* If the PR is unfinished, add `[WIP]` prefix to the pull request title and add the `do-not-merge` label.
* Include in the title the category of this pull request. The categories are: `[WIP]`, `[Feature]` or `[Fix]`. If the category is unknown use `[WIP]`, once the PR is ready to be merged replace `[WIP]` for the right category.
* Always assign label(s) to your PR.

**What this PR does / Why we need it:**
This PR allow the filtering of the Terraform logs. 
As stated in the issues #8, #17 and #18 the use of Terranova cause a flood of Terraform logs that are barely impossible to read and understand. The goal of this PR is to intercept the TF logs and print the entries the developer wants to print.

**References:**
List of issues this PR fixes using the prefix `Fixes`, `Closes` or `Resolves` and the issue number or link. Format Examples: `Fixes #<issue number>`, `Closes <link of issue>` or `Resolves #<issue number>`.
The issues will be closed automatically when this pull request is merged. Example:
 - Fixes #8 
 - Fixes #17 
 - Fixes #18  

**How I did it:**
The Terranova Log Middleware is going to intercept every output from the standard Log instance (`log`) used by Terraform and Providers to print status, information, debugs, warnings and errors. The user optionally provide a logger which implement the methods from the Terranova Logger interface to print the log entries they want and in the format they want. If the logger is not provided, Terranova will use a default one that only prints the Info, Warn and Errors log entries.

**How to verify it:**
There are some (incomplete) unit test that can be used to verify this PR. My intention is to complete the unit test in another PR that will include also all the unit test for Terranova.
There are also a set of examples in the [Terranova Examples](https://github.com/johandry/terranova-examples/tree/master/custom-logs) repository that can be used to test and know how to use this new feature

**Description for the changelog:**
- Use of Log Middleware and Logger to filter the Terraform logs

**How to use this feature:**
The [Terranova Examples](https://github.com/johandry/terranova-examples/tree/master/custom-logs) repository contain several sample code to know how to use this feature. 
A documentation will be included in this PR to explain how to use it.
